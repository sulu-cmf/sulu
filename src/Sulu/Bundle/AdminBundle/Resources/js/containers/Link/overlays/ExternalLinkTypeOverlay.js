// @flow
import React, {Fragment} from 'react';
import {observer} from 'mobx-react';
import {action, observable} from 'mobx';
import Dialog from '../../../components/Dialog';
import Form from '../../../components/Form';
import Input from '../../../components/Input';
import SingleSelect from '../../../components/SingleSelect';
import TextArea from '../../../components/TextArea';
import Url from '../../../components/Url';
import {translate} from '../../../utils';
import type {LinkTypeOverlayProps} from '../types';

@observer
class ExternalLinkTypeOverlay extends React.Component<LinkTypeOverlayProps> {
    @observable protocol: ?string = undefined;
    @observable href: ?string = undefined;
    @observable mailSubject: ?string = undefined;
    @observable mailBody: ?string = undefined;

    constructor(props: LinkTypeOverlayProps) {
        super(props);

        this.updateUrl();
    }

    @action componentDidUpdate(prevProps: LinkTypeOverlayProps) {
        if (prevProps.open === false && this.props.open === true) {
            this.updateUrl();
        }
    }

    updateUrl() {
        const {href} = this.props;

        if (!href) {
            this.href = undefined;
            return;
        }

        if (href.startsWith('mailto:')) {
            const urlParts = String(href).split('?');
            const urlParameters = new URLSearchParams(urlParts[1]);
            const mailSubject = urlParameters.get('subject');
            const mailBody = urlParameters.get('body');

            this.href = urlParts[0];
            this.mailSubject = mailSubject ? mailSubject : undefined;
            this.mailBody = mailBody ? mailBody : undefined;

            return;
        }

        this.href = href;
        this.mailSubject = undefined;
        this.mailBody = undefined;
    }

    callUrlChange = () => {
        const {onTargetChange, onHrefChange} = this.props;
        const {mailBody, mailSubject, href} = this;

        if (!href) {
            onHrefChange(undefined);
            return;
        }

        const urlParameters = new URLSearchParams();

        if (href.startsWith('mailto:')) {
            if (onTargetChange) {
                onTargetChange('_self');
            }

            if (mailSubject) {
                urlParameters.set('subject', mailSubject);
            }

            if (mailBody) {
                urlParameters.set('body', mailBody);
            }
        }

        onHrefChange(
            href + (
                Array.from(urlParameters).length > 0
                    // Replacing value is required, because Apple Mail does not seem to understand it otherwise
                    ? '?' + urlParameters.toString().replace(/\+/g, '%20')
                    : ''
            )
        );
    };

    handleUrlBlur = this.callUrlChange;

    @action handleHrefChange = (href: ?string) => {
        this.href = href;
    };

    handleMailSubjectBlur = this.callUrlChange;

    @action handleProtocolChange = (protocol: ?string) => {
        this.protocol = protocol;
    };

    @action handleMailSubjectChange = (mailSubject: ?string) => {
        this.mailSubject = mailSubject;
    };

    handleMailBodyBlur = this.callUrlChange;

    @action handleMailBodyChange = (mailBody: ?string) => {
        this.mailBody = mailBody;
    };

    render() {
        const {
            onCancel,
            onConfirm,
            onTargetChange,
            onTitleChange,
            open,
            target,
            title,
            href,
        } = this.props;

        return (
            <Dialog
                cancelText={translate('sulu_admin.cancel')}
                confirmDisabled={!href}
                confirmText={translate('sulu_admin.confirm')}
                onCancel={onCancel}
                onConfirm={onConfirm}
                open={open}
                title={translate('sulu_admin.link')}
            >
                <Form>
                    <Form.Field label={translate('sulu_admin.link_url')} required={true}>
                        <Url
                            defaultProtocol="https://"
                            onBlur={this.handleUrlBlur}
                            onChange={this.handleHrefChange}
                            onProtocolChange={this.handleProtocolChange}
                            valid={true}
                            value={this.href}
                        />
                    </Form.Field>

                    {this.protocol && this.protocol !== 'mailto:' && onTargetChange &&
                        <Form.Field label={translate('sulu_admin.link_target')} required={true}>
                            <SingleSelect onChange={onTargetChange} value={target}>
                                <SingleSelect.Option value="_blank">_blank</SingleSelect.Option>
                                <SingleSelect.Option value="_self">_self</SingleSelect.Option>
                                <SingleSelect.Option value="_parent">_parent</SingleSelect.Option>
                                <SingleSelect.Option value="_top">_top</SingleSelect.Option>
                            </SingleSelect>
                        </Form.Field>
                    }

                    {this.protocol && this.protocol === 'mailto:' &&
                        <Fragment>
                            <Form.Field label={translate('sulu_admin.mail_subject')}>
                                <Input
                                    onBlur={this.handleMailSubjectBlur}
                                    onChange={this.handleMailSubjectChange}
                                    value={this.mailSubject}
                                />
                            </Form.Field>
                            <Form.Field label={translate('sulu_admin.mail_body')}>
                                <TextArea
                                    onBlur={this.handleMailBodyBlur}
                                    onChange={this.handleMailBodyChange}
                                    value={this.mailBody}
                                />
                            </Form.Field>
                        </Fragment>
                    }

                    {onTitleChange &&
                        <Form.Field label={translate('sulu_admin.link_title')}>
                            <Input onChange={onTitleChange} value={title} />
                        </Form.Field>
                    }
                </Form>
            </Dialog>
        );
    }
}

export default ExternalLinkTypeOverlay;
