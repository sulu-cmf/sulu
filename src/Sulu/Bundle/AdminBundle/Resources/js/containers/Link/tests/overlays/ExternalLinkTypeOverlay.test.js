// @flow
import React from 'react';
import {mount, shallow} from 'enzyme';
import ExternalLinkTypeOverlay from '../../overlays/ExternalLinkTypeOverlay';

jest.mock('../../../../utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

test('Render overlay with an undefined URL', () => {
    const externalLinkOverlay = mount(
        <ExternalLinkTypeOverlay
            href={undefined}
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={jest.fn()}
            onTargetChange={jest.fn()}
            onTitleChange={jest.fn()}
            open={true}
            target={undefined}
            title={undefined}
        />
    );

    expect(externalLinkOverlay.find('Form').render()).toMatchSnapshot();
});

test('Render overlay with mailto URL', () => {
    const externalLinkOverlay = mount(
        <ExternalLinkTypeOverlay
            href="mailto:test@example.org?subject=Subject&body=Body"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={jest.fn()}
            onTargetChange={jest.fn()}
            onTitleChange={jest.fn()}
            open={true}
            target={undefined}
            title={undefined}
        />
    );

    expect(externalLinkOverlay.find('Form').render()).toMatchSnapshot();
});

test('Render overlay with a URL', () => {
    const externalLinkOverlay = mount(
        <ExternalLinkTypeOverlay
            href="http://www.sulu.io"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={jest.fn()}
            onTargetChange={jest.fn()}
            onTitleChange={jest.fn()}
            open={true}
            target={undefined}
            title={undefined}
        />
    );

    expect(externalLinkOverlay.find('Form').render()).toMatchSnapshot();
});

test('Pass correct props to Dialog', () => {
    const cancelSpy = jest.fn();
    const confirmSpy = jest.fn();

    const externalLinkOverlay = shallow(
        <ExternalLinkTypeOverlay
            href={undefined}
            onCancel={cancelSpy}
            onConfirm={confirmSpy}
            onHrefChange={jest.fn()}
            onTargetChange={jest.fn()}
            onTitleChange={jest.fn()}
            open={false}
            target={undefined}
            title={undefined}
        />
    );

    expect(externalLinkOverlay.find('Dialog').prop('onCancel')).toEqual(cancelSpy);
    expect(externalLinkOverlay.find('Dialog').prop('onConfirm')).toEqual(confirmSpy);
    expect(externalLinkOverlay.find('Dialog').prop('open')).toEqual(false);
});

test('Display given URL with query parameters in href input', () => {
    const targetChangeSpy = jest.fn();
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = shallow(
        <ExternalLinkTypeOverlay
            href="http://www.sulu.io/contact-us?param=value-1"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={targetChangeSpy}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    expect(externalLinkOverlay.find('Url').prop('value')).toEqual('http://www.sulu.io/contact-us?param=value-1');
});

test('Do not call onHrefChange handler if input did not loose focus', () => {
    const targetChangeSpy = jest.fn();
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = shallow(
        <ExternalLinkTypeOverlay
            href={undefined}
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={targetChangeSpy}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    externalLinkOverlay.find('Url').prop('onChange')('http://www.sulu.io');
    expect(urlChangeSpy).not.toBeCalled();
});

test('Fields should change immediately after protocol was changed', () => {
    const targetChangeSpy = jest.fn();
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = mount(
        <ExternalLinkTypeOverlay
            href={undefined}
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={targetChangeSpy}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    expect(externalLinkOverlay.find('Field[label="sulu_admin.link_target"]')).toHaveLength(1);
    expect(externalLinkOverlay.find('Field[label="sulu_admin.mail_subject"]')).toHaveLength(0);
    expect(externalLinkOverlay.find('Field[label="sulu_admin.mail_body"]')).toHaveLength(0);

    externalLinkOverlay.find('Url').prop('onProtocolChange')('mailto:');

    externalLinkOverlay.update();
    expect(externalLinkOverlay.find('Field[label="sulu_admin.link_target"]')).toHaveLength(0);
    expect(externalLinkOverlay.find('Field[label="sulu_admin.mail_subject"]')).toHaveLength(1);
    expect(externalLinkOverlay.find('Field[label="sulu_admin.mail_body"]')).toHaveLength(1);
});

test('Call onHrefChange with URL that includes mail subject and mail body for mailto links', () => {
    const targetChangeSpy = jest.fn();
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = shallow(
        <ExternalLinkTypeOverlay
            href="mailto:bla@example.org"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={targetChangeSpy}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    externalLinkOverlay.find('Url').prop('onChange')('mailto:test@example.org');
    externalLinkOverlay.find('Url').prop('onProtocolChange')('mailto:');
    expect(urlChangeSpy).not.toBeCalledWith('mailto:test@example.org');
    externalLinkOverlay.find('Url').prop('onBlur')();
    expect(urlChangeSpy).toBeCalledWith('mailto:test@example.org');
    expect(targetChangeSpy).toBeCalledWith('_self');

    externalLinkOverlay.update();
    externalLinkOverlay.find('Field[label="sulu_admin.mail_subject"] Input').prop('onChange')('Subject Line');
    expect(urlChangeSpy).not.toBeCalledWith('mailto:test@example.org?subject=Subject%20Line');
    externalLinkOverlay.find('Field[label="sulu_admin.mail_subject"] Input').prop('onBlur')();
    expect(urlChangeSpy).toBeCalledWith('mailto:test@example.org?subject=Subject%20Line');

    externalLinkOverlay.update();
    externalLinkOverlay.find('Field[label="sulu_admin.mail_body"] TextArea').prop('onChange')('Body Text');
    expect(urlChangeSpy).not.toBeCalledWith('mailto:test@example.org?subject=Subject%20Line&body=Body%20Text');
    externalLinkOverlay.find('Field[label="sulu_admin.mail_body"] TextArea').prop('onBlur')();
    expect(urlChangeSpy).toBeCalledWith('mailto:test@example.org?subject=Subject%20Line&body=Body%20Text');
});

test('Should not include mail subject and body in URL after switching to another protocol', () => {
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = mount(
        <ExternalLinkTypeOverlay
            href="mailto:test@example.org?subject=Subject&body=Body%20Text"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={jest.fn()}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    externalLinkOverlay.find('Url').prop('onBlur')();
    expect(urlChangeSpy).toHaveBeenNthCalledWith(1, 'mailto:test@example.org?subject=Subject&body=Body%20Text');

    externalLinkOverlay.find('Url').find('SingleSelect').prop('onChange')('https://');
    expect(urlChangeSpy).toHaveBeenNthCalledWith(2, 'https://test@example.org');
});

test('Reset target to self when a mailto link is entered', () => {
    const targetChangeSpy = jest.fn();
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = shallow(
        <ExternalLinkTypeOverlay
            href="http://www.sulu.io"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={targetChangeSpy}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    externalLinkOverlay.find('Url').prop('onChange')('mailto:test@example.org');
    externalLinkOverlay.find('Url').prop('onBlur')();
    expect(urlChangeSpy).toBeCalledWith('mailto:test@example.org');
    expect(targetChangeSpy).toBeCalledWith('_self');
});

test('Should not reset target to self when a non-mail URL is entered', () => {
    const targetChangeSpy = jest.fn();
    const urlChangeSpy = jest.fn();

    const externalLinkOverlay = shallow(
        <ExternalLinkTypeOverlay
            href="http://www.sulu.io"
            onCancel={jest.fn()}
            onConfirm={jest.fn()}
            onHrefChange={urlChangeSpy}
            onTargetChange={targetChangeSpy}
            onTitleChange={jest.fn()}
            open={true}
            target="_blank"
            title={undefined}
        />
    );

    externalLinkOverlay.find('Url').prop('onChange')('http://sulu.io');
    externalLinkOverlay.find('Url').prop('onBlur')();
    expect(urlChangeSpy).toBeCalledWith('http://sulu.io');
    expect(targetChangeSpy).not.toBeCalled();
});
