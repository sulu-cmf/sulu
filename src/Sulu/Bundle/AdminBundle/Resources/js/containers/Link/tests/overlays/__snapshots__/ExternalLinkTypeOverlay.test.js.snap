// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render overlay with a URL 1`] = `
<div
  class="grid grid"
>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_url *
      </label>
      <div
        class="url"
      >
        <div
          class="protocols"
        >
          <div
            class="select"
            role="none"
          >
            <button
              class="displayValue flat"
              type="button"
            >
              <div
                aria-label="http://"
                class="croppedText"
                title="http://"
              >
                <div
                  aria-hidden="true"
                  class="front"
                >
                  http
                </div>
                <div
                  aria-hidden="true"
                  class="back"
                >
                  <span>
                    ://
                  </span>
                </div>
                <div
                  class="whole"
                >
                  http://
                </div>
              </div>
              <span
                aria-label="su-angle-down"
                class="su-angle-down toggle"
              />
            </button>
          </div>
        </div>
        <input
          type="text"
          value="www.sulu.io"
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_target *
      </label>
      <div
        class="select"
        role="none"
      >
        <button
          class="displayValue default"
          type="button"
        >
          <div
            aria-label="sulu_admin.please_choose"
            class="croppedText"
            title="sulu_admin.please_choose"
          >
            <div
              aria-hidden="true"
              class="front"
            >
              sulu_admin.p
            </div>
            <div
              aria-hidden="true"
              class="back"
            >
              <span>
                lease_choose
              </span>
            </div>
            <div
              class="whole"
            >
              sulu_admin.please_choose
            </div>
          </div>
          <span
            aria-label="su-angle-down"
            class="su-angle-down toggle"
          />
        </button>
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_title
      </label>
      <div
        class="input default left"
      >
        <input
          type="text"
          value=""
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
        tabindex="-1"
      >
        <span
          class="switch toggler"
        >
          <input
            type="checkbox"
            value=""
          />
          <span />
        </span>
        <div>
          sulu_admin.no_follow
        </div>
      </label>
      <div
        class="errorLabel"
      />
    </div>
  </div>
</div>
`;

exports[`Render overlay with an undefined URL 1`] = `
<div
  class="grid grid"
>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_url *
      </label>
      <div
        class="url"
      >
        <div
          class="protocols"
        >
          <div
            class="select"
            role="none"
          >
            <button
              class="displayValue flat"
              type="button"
            >
              <div
                aria-label="https://"
                class="croppedText"
                title="https://"
              >
                <div
                  aria-hidden="true"
                  class="front"
                >
                  http
                </div>
                <div
                  aria-hidden="true"
                  class="back"
                >
                  <span>
                    s://
                  </span>
                </div>
                <div
                  class="whole"
                >
                  https://
                </div>
              </div>
              <span
                aria-label="su-angle-down"
                class="su-angle-down toggle"
              />
            </button>
          </div>
        </div>
        <input
          type="text"
          value=""
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_target *
      </label>
      <div
        class="select"
        role="none"
      >
        <button
          class="displayValue default"
          type="button"
        >
          <div
            aria-label="sulu_admin.please_choose"
            class="croppedText"
            title="sulu_admin.please_choose"
          >
            <div
              aria-hidden="true"
              class="front"
            >
              sulu_admin.p
            </div>
            <div
              aria-hidden="true"
              class="back"
            >
              <span>
                lease_choose
              </span>
            </div>
            <div
              class="whole"
            >
              sulu_admin.please_choose
            </div>
          </div>
          <span
            aria-label="su-angle-down"
            class="su-angle-down toggle"
          />
        </button>
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_title
      </label>
      <div
        class="input default left"
      >
        <input
          type="text"
          value=""
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
        tabindex="-1"
      >
        <span
          class="switch toggler"
        >
          <input
            type="checkbox"
            value=""
          />
          <span />
        </span>
        <div>
          sulu_admin.no_follow
        </div>
      </label>
      <div
        class="errorLabel"
      />
    </div>
  </div>
</div>
`;

exports[`Render overlay with mailto URL 1`] = `
<div
  class="grid grid"
>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_url *
      </label>
      <div
        class="url"
      >
        <div
          class="protocols"
        >
          <div
            class="select"
            role="none"
          >
            <button
              class="displayValue flat"
              type="button"
            >
              <div
                aria-label="mailto:"
                class="croppedText"
                title="mailto:"
              >
                <div
                  aria-hidden="true"
                  class="front"
                >
                  mail
                </div>
                <div
                  aria-hidden="true"
                  class="back"
                >
                  <span>
                    to:
                  </span>
                </div>
                <div
                  class="whole"
                >
                  mailto:
                </div>
              </div>
              <span
                aria-label="su-angle-down"
                class="su-angle-down toggle"
              />
            </button>
          </div>
        </div>
        <input
          type="text"
          value="test@example.org"
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.mail_subject
      </label>
      <div
        class="input default left"
      >
        <input
          type="text"
          value="Subject"
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.mail_body
      </label>
      <textarea
        class="textArea"
      >
        Body
      </textarea>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        sulu_admin.link_title
      </label>
      <div
        class="input default left"
      >
        <input
          type="text"
          value=""
        />
      </div>
      <div
        class="errorLabel"
      />
    </div>
  </div>
  <div
    class="item gridItem colSpan colSpan-12 space-before-0 space-after-0"
  >
    <div
      class="field"
    >
      <label
        class="label"
        tabindex="-1"
      >
        <span
          class="switch toggler"
        >
          <input
            type="checkbox"
            value=""
          />
          <span />
        </span>
        <div>
          sulu_admin.no_follow
        </div>
      </label>
      <div
        class="errorLabel"
      />
    </div>
  </div>
</div>
`;
